File: Backend/dependency/notifications/src/main/java/com/mhp/coding/challenges/dependency/notifications/InquiryEventListener.java
package com.mhp.coding.challenges.dependency.notifications;

import com.mhp.coding.challenges.dependency.inquiry.Inquiry;
import lombok.RequiredArgsConstructor;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.context.event.EventListener;
import org.springframework.stereotype.Component;

@Component
@RequiredArgsConstructor
public class InquiryEventListener {

    private static final Logger LOG = LoggerFactory.getLogger(InquiryEventListener.class);

    private final EmailHandler emailHandler;
    private final PushNotificationHandler pushNotificationHandler;

    @EventListener
    public void onApplicationEvent(Inquiry inquiry) {
        LOG.info("received inquiry {}", inquiry);
        emailHandler.sendEmail(inquiry);
        pushNotificationHandler.sendNotification(inquiry);
    }
}

File: Backend/dependency-kotlin/notifications/src/main/kotlin/com/mhp/coding/challenges/dependency/notifications/InquiryEventListener.kt
package com.mhp.coding.challenges.dependency.notifications

import com.mhp.coding.challenges.dependency.inquiry.InquiryEvent
import mu.KotlinLogging
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.event.EventListener
import org.springframework.stereotype.Component

private val logger = KotlinLogging.logger {}

@Component
class InquiryEventListener(
    @Autowired val emailHandler: EmailHandler,
    @Autowired val pushNotificationHandler: PushNotificationHandler
) {

    @EventListener
    fun onApplicationEvent(event: InquiryEvent) {
        logger.info { "Received InquiryEvent: $event" }
        emailHandler.sendEmail(event.inquiry)
        pushNotificationHandler.sendNotification(event.inquiry)
    }
}
